### ISSUE TITLE
Clang accepts invalid conversion of lambda to std::function 

### ISSUE THREAD

**Issue by ranaanoop** [llvm/llvm-project#85495]
The following program compiles with clang without any diagnostic but rejected by gcc: [Demo](https://godbolt.org/z/79hrndn3E)

```
#include <functional>
#include <iostream>
#include <string>
int func() 
{ 
    return 5; 
}
int main()  
{
    
    std::function<const int&()> getter1 = 
                                        [] { return func(); }; //gcc rejects but clang accepts  
    getter1();
}
```

**Comment by shafik** [comment]
If we drop the `const` from the `std::function` signature: https://godbolt.org/z/vPa9hWKTT

then implementations agree. It sure seems like a clang bug but maybe I am missing something.

CC @zygoloid

**Comment by zygoloid** [comment]
Seems like a stdlib issue; is this libc++ vs libstdc++, or is it really GCC vs Clang for the same stdlib?

**Comment by zygoloid** [comment]
This is the rule that's not being properly implemented here: https://eel.is/c++draft/function.objects#func.require-2.sentence-2

I think that was added fairly recently; it's also worth checking the compilers are targeting the same standard.

**Comment by RichardLuo0** [comment]
I think this needs more attention. The lambda that returns a value may be accidentally converted into std::function which returns a reference. This may create dangerous dangling references unexpectedly.
The compiler should throw an error or at least a warning that this might not be the desired conversion.
